### ISSUE TITLE
Allow to inject JPA repositories for different datasources

### ISSUE THREAD

**Issue by esocode** [micronaut-projects/micronaut-data#368]
I have an application that needs to use the same JPA repository on different datasources. At first I tried to inject it with `@PersistenceContext(name = "<datasource>")` like it works for a transaction manager but without success. I just found out that the `@Repository` annotation can be used with a datasource name. But in my opinion that is the wrong place for this declaration. It requires to define multiple repository interfaces which have exactly the same structure. Possibly that could be solved with inheritance (haven't tried) but it still contradicts the logical structure as that would be a declaration just for runtime purposes. It also makes it difficult to define repositories in libraries.

Wouldn't it be possible to use the `@PersistenceContext` annotation for repository injection too? 

### Expected Behaviour

Inject JPA repository variables for different (non-default) datasources.

### Actual Behaviour

The datasource can only be controlled by `@Repository` annotation, requiring to define multiple identical repository interfaces.

### Environment Information

- **Micronaut Version:** 1.3.0.M2
- **micronaut-data Version:** 1.0.0.M5

**Comment by jeffbrown** [comment]
For reference: https://stackoverflow.com/questions/59737148/how-to-inject-a-jpa-repository-for-multiple-datasources-in-micronaut-data

**Comment by graemerocher** [comment]
This will be feasable when we rebase Micronaut Data on Micronaut 2.0

**Comment by esocode** [comment]
Any news about this?

**Comment by HimmelSpark** [comment]
Hi there, any news?

**Comment by Hc747** [comment]
Bump

**Comment by Maritims** [comment]
Has there been any development in this matter?

**Comment by dogcow** [comment]
This feature would be very helpful to me too.  @esocode, have you found a workaround?

**Comment by esocode** [comment]
No. I think there is currently no easy way, you have to duplicate code to access multiple repositories.

**Comment by dogcow** [comment]
@esocode , I'm hacking around this by getting EntityManager beans & TransactionManager beans from the application context for each datasource.  I did something along these lines (not actual code):

```
public WidgetRepository(ApplicationContext context)
{
    for (WidgetDatabase wd : widgetDatabaseList)
    {
        EntityManager entityManager = context.getBean(EntityManager.class, Qualifier.byName(wd.datasourceName));
        SynchronousTransactionManager<?> transactionManager = context.getBean(SynchronousTransactionManager.class, Qualifiers.byName(wd.datasourceName));
        datasourceMap.put(wd.datasourceName, new EntityAndTransactionManagers(shardEntityManager, shardTransactionManager));
    }
}

public Optional<Widget> findByNameQueryOnDatasource(@NonNull String widgetName, String datasourceName)
{
    EntityAndTransactionManagers managers = datasourceMap.get(datasourceName);
    return mangers.getTransactionManager().executeRead(status -> {
            return managers.getEntityManager().createQuery(query, Widget.class)
            .setParameter("widgetName", widgetName)
            .getResultsStream()
            .findFirst();
}
    
```

I still need to fully test this, but it seems to be working at first glance.

**Comment by Maritims** [comment]
How's the experiment coming along @dogcow ?

**Comment by Maritims** [comment]
@dogcow Would you be willing to share the actual code? This might never make it into Micronaut for all we know.

**Comment by dogcow** [comment]
I could flesh out my example code above.

**Comment by Maritims** [comment]
@dogcow That'd be great!

**Comment by dogcow** [comment]
ok, I've updated my previous example to be a little more complete.  I haven't tested the example code, but it should be enough to get you "over the hump" and able to use a single "repository" class to CRUD multiple datasources.  I'm sure there's a way to avoid the "dbhosts" configuration & pull the list straight from the "datasources" configuration, but I haven't dug into it yet.

**Comment by dstepanov** [comment]
I have a PR which should allow it https://github.com/micronaut-projects/micronaut-data/pull/1608
